docs/advanced-guides/testing/testing-with-react-testing-library.md

```diff
@@ -1,13 +1,153 @@
+  fireEvent.click(aboutLink);
+
+  expect(window.location.pathname).toBe('/about');
+
+  const homeLink = screen.getByText('Home');
+  fireEvent.click(homeLink);
+
+  expect(window.location.pathname).toBe('/');
+});
+```
+
+This test will initially fail if it is run immediately after the previous test. This is because the history stack still thinks we are on the about page. To fix this, we can use Jest's `beforeEach` function to replace our history state using: `window.history.replaceState({}, '', '/');`. This makes `react-router` think that we are on the home page when we start each test. Now, all three tests should be passing when run together.
+
+Below is the full file with all of the tests together:
+
+```jsx
+import React from 'react';
+import { render, fireEvent, screen } from '@testing-library/react';
+import App from './App';
+import { BrowserRouter as Router } from 'react-router-dom';
+
+beforeEach(() => {
+  window.history.replaceState({}, '', '/');
+});
+
+test('<App /> renders successfully', () => {
+  render(<App />, { wrapper: Router });
+  const header = screen.getByText('Welcome to React Router!');
+  expect(header).toBeInTheDocument();
+});
+
+it('navigates to /about', () => {
+  render(<App />, { wrapper: Router });
+
+  const aboutLink = screen.getByText('About');
+  fireEvent.click(aboutLink);
+
+  expect(window.location.pathname).toBe('/about');
+});
+
+it('navigates to /about and back to /', () => {
+  render(<App />, { wrapper: Router });
+
+  const aboutLink = screen.getByText('About');
+  fireEvent.click(aboutLink);
+
+  expect(window.location.pathname).toBe('/about');
+
+  const homeLink = screen.getByText('Home');
+  fireEvent.click(homeLink);
+
+  expect(window.location.pathname).toBe('/');
+});
+```
+
+## Testing Routes and Redirects
+
+Let's say that we've built a redirect component that deals with re-routing the user from `/the-old-thing` to `/the-new-thing`
+
+`RedirectHandler.js`
+```jsx
+import React from 'react'
+import {Route, Switch, Redirect} from 'react-router-dom'
+
+const RedirectHandler = props => (
+  <Switch>
+    <Redirect from="/the-old-thing" to="the-new-thing">
+  </Switch>
+)
+
+export default RedirectHandler
+```
+
+And here we have one method of testing that the redirect is behaving as expected.
+
+`RedirectHandler.test.js`
+```jsx
+import React from 'react'
+import {MemoryRouter, Route} from 'react-router-dom'
+import { render, fireEvent, screen } from '@testing-library/react';
+import RedirectHandler from './RedirectHandler'
+
+it('redirects /the-old-thing to /the-new-thing', () => {
+  render(
+    <MemoryRouter initialEntries={["/the-old-thing"]}>
+      <Route component={RedirectHandler}>
+    </MemoryRouter>
+  )
+
+  expect(window.location.pathname).toBe("/the-new-thing")
+})
+
+```
+
+This test initializes the `MemoryRouter` with an initial path of `/the-old-thing` and verifies that the `RedirectHandler` correctly changes the path to `/the-new-thing`
```

Comment:
Will this actually happen with a MemoryRouter? I don't think it updates the location. But you could read from the history object.

Reply:
Just checked the docs, you're correct that `MemoryRouter` won't update `window.location`.

I'm not sure how to set the equivalent of this

```jsx
initialEntries={["/the-old-thing"]}
```
using `BrowserRouter`, in order to set up for the redirect. 

Is it acceptable to directly access `history.location` for testing purposes?

Reply:
I don't really know why `initialEntries` exists (maybe to restore across sessions?). Pushing history and waiting for redirect to result in the new URL seems more realistic to me for web use cases. Again, there are limitations to both approaches. `window.location.href = ` doesn't work in JSDOM, for example.